### DISCUSSION TITLE
Letter spacing and bullet points

### DISCUSSION THREAD

**Discussion by Mat2001mat** [QuestPDF/QuestPDF#165] [Q&A] [answered]
Firstly, thank you. This library is fantastic and very quick to pick up.

A couple of things I'm not sure about. Is it possible to use letter spacing? For example, the Material System uses different letter spacing for its font scales. https://material.io/design/typography/the-type-system.html#type-scale

Secondly, is there a way to use bullet points?

Many thanks

**Comment by MarcinZiabek** [comment, marked as answer]
Hi 😁 Let me answer your questions:

> A couple of things I'm not sure about. Is it possible to use letter spacing? For example, the Material System uses different letter spacing for its font scales. https://material.io/design/typography/the-type-system.html#type-scale

Right now the current text rendering engine does not support custom letter spacing. This option is not supported by SkiaSharp and therefore cannot be easily surfaced to QuestPDF API. However, I am slowly investigating implementation of more advanced algorithm. Something that support text shaping, RTL, text justification and possibly letter spacing. This is a long process though and I don't expect this in next release.

> Secondly, is there a way to use bullet points?

There is no element that provides this functionality out of the box. Please take a look on how to create such layout using already existing elements:

```csharp
.Column(column =>
{
    foreach (var _ in Enumerable.Range(0, 8))
    {
        column.Item().Row(row =>
        {
            row.Spacing(5);
            row.AutoItem().Text("-"); // text or image
            row.RelativeItem().Text(Placeholders.Sentence());
        });
    }
});
```

The code outputs:
<img src="https://user-images.githubusercontent.com/9263853/161645840-5c218a51-12f3-448e-90f7-38fc063770c6.png" width="400">

If you feel that this functionality should be covered by a new structural block and be available in the API out of the box, please suggest how it should look like 😁 All ideas are always welcome!

**Reply by Mat2001mat** [reply]
Hi Marcin,

Appreciate your detailed reply. Once again, thank you for creating this library. It feels very similar to using something like Flutter and I've been able to create very accurate PDF's to my original designs. 

Ah, okay. That makes sense and why it's not easy to do so. I did experiment with the Scale option, but it didn't quite look right sadly. I'm not sure what the level of interest is in more advanced text rendering and whether it's worth the time investment, but it's definitely a nice to have.

Great, that works nicely.  In my case, I used ASCII to generate the bullet point icon and it looked clean. I don't think it needs an out of the box solution.

`        col.Item()
                                .PaddingTop(Theme.BulletSpace)
                                .Text(txt =>
                                {
                                    txt.Span($"• ")
                                        .Style(Typography.Body1)
                                        .Bold();
                                    txt.Span($"{skill}")
                                        .Style(Typography.Body1);
                                });`

All the best,

Matt

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
